dll/cpl/sysdm/general.c

```diff
@@ -417,14 +431,393 @@ static VOID SetProcSpeed(HWND hwnd, HKEY hKey, LPTSTR Value, UINT uID)
+        { L"Super Server", TRUE },            // some Supermicro server boards
+    };
+    BOOL bMatch;
+    UINT i;
+
+    for (i = 0; i < _countof(Vendors); i++)
+    {
+        if (Vendors[i].bCaseSensitive)
+        {
+            bMatch = !wcscmp(ven, Vendors[i].pwName);
+        }
+        else
+        {
+            bMatch = !wcsicmp(ven, Vendors[i].pwName);
+        }
+        if (bMatch)
+        {
+            return TRUE;
+        }
+    }
+
+    for (i = 0; i < _countof(Devices); i++)
+    {
+        if (Devices[i].bCaseSensitive)
+        {
+            bMatch = !wcscmp(dev, Devices[i].pwName);
+        }
+        else
+        {
+            bMatch = !wcsicmp(dev, Devices[i].pwName);
+        }
+        if (bMatch)
+        {
+            return TRUE;
+        }
+    }
+    return FALSE;
+}
+
+static BOOL GetSystemName(PWSTR pBuf, SIZE_T cchBuf)
+{
+    static const VENDOR_LONG_NAME LongNames[] =
+    {
+        { L"ASUSTeK", L"ASUS" },
+        { L"First International Computer", L"FIC" },
+        { L"Hewlett-Packard", L"HP" },
+        { L"MICRO-STAR", L"MSI" },
+        { L"SGI.COM", L"SGI" },
+        { L"Silicon Graphics International", L"SGI" },
+    };
+    static const PCWSTR RedundantWords[] =
+    {
+        L"Corporation",
+        L"Computer",
+        L"Computers",
+        L"Group",
+        L"Microsystems",
+        L"Electronics",
+        L"Electric",
+        L"Software",
+        L"International",
+        L"Industrial",
+        L"Technology",
+        L"Tecohnology", // on purpose (some Gigabyte boards)
+        L"Inc.",
+        L"Inc",
+        L"Co.",
+        L"Co",
+        L"Ltd.",
+        L"Ltd",
+        L"GmbH",
+        L"INT'L",
+    };
+    PVOID SMBiosBuf;
+    PCHAR DmiStrings[ID_STRINGS_MAX] = { 0 };
+    WCHAR ven[512], dev[512];
+    UINT i;
+
+    SMBiosBuf = LoadSMBiosData(DmiStrings);
+    if (!SMBiosBuf)
+    {
+        return FALSE;
+    }
+
+    GetSMBiosStringW(DmiStrings[SYS_VENDOR], ven, _countof(ven), TRUE);
+    GetSMBiosStringW(DmiStrings[SYS_PRODUCT], dev, _countof(dev), TRUE);
+
+    if (wcslen(dev) == 0 ||
+        !wcscmp(dev, ven) ||
+        isGenericSystemName(ven, dev))
+    {
+        // system strings are unusable, use board strings
+        GetSMBiosStringW(DmiStrings[BOARD_VENDOR], ven, _countof(ven), TRUE);
+        GetSMBiosStringW(DmiStrings[BOARD_NAME], dev, _countof(dev), TRUE);
+
+        if (wcslen(ven) == 0 && wcslen(dev) == 0)
+        {
+            // board strings are empty, use BIOS vendor string
+            GetSMBiosStringW(DmiStrings[BIOS_VENDOR], ven, _countof(ven), TRUE);
+        }
+    }
+
+    // make vendor strings shorter
+    for (i = 0; i < _countof(LongNames); i++)
+    {
+        wcsrep(ven, LongNames[i].pwLongName, LongNames[i].pwShortName);
+    }
+
+    // remove redundant words
+    for (i = 0; i < _countof(RedundantWords); i++)
+    {
+        wcsrep(ven, RedundantWords[i], L"");
```

Comment:
What if the word is part of another word?
Like, `Inc` -> `Incorporated` -> `corporated`

Reply:
`wcsrep` replaces full word by adding a space at begin and end of string.